ddl/ddl_tiflash_api.go

```diff
@@ -0,0 +1,512 @@
+				tblInfo.TiFlashReplica.Count, tblInfo.TiFlashReplica.LocationLabels, tblInfo.TiFlashReplica.IsPartitionAvailable(p.ID), false})
+		}
+		// partitions that in adding mid-state
+		for _, p := range pi.AddingDefinitions {
+			logutil.BgLogger().Debug(fmt.Sprintf("Table %v has partition %v\n", tblInfo.ID, p.ID))
+			*tableList = append(*tableList, TiFlashReplicaStatus{p.ID, tblInfo.TiFlashReplica.Count, tblInfo.TiFlashReplica.LocationLabels, tblInfo.TiFlashReplica.IsPartitionAvailable(p.ID), true})
+		}
+	} else {
+		logutil.BgLogger().Debug(fmt.Sprintf("Table %v has no partition\n", tblInfo.ID))
+		*tableList = append(*tableList, TiFlashReplicaStatus{tblInfo.ID, tblInfo.TiFlashReplica.Count, tblInfo.TiFlashReplica.LocationLabels, tblInfo.TiFlashReplica.Available, false})
+	}
+}
+
+// UpdateTiFlashHTTPAddress report TiFlash's StatusAddress's port to Pd's etcd.
+func (d *ddl) UpdateTiFlashHTTPAddress(store *helper.StoreStat) error {
+	addrAndPort := strings.Split(store.Store.StatusAddress, ":")
+	if len(addrAndPort) < 2 {
+		return errors.New("Can't get TiFlash Address from PD")
+	}
+	httpAddr, err := getTiflashHTTPAddr(addrAndPort[0], store.Store.StatusAddress)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	// Report to pd
+	key := fmt.Sprintf("/tiflash/cluster/http_port/%v", store.Store.Address)
+	if d.etcdCli == nil {
+		return errors.New("no etcdCli in ddl")
+	}
+	origin := ""
+	resp, err := d.etcdCli.Get(d.ctx, key)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	// Try to update.
+	for _, kv := range resp.Kvs {
+		if string(kv.Key) == key {
+			origin = string(kv.Value)
+			break
+		}
+	}
+	if origin != httpAddr {
+		logutil.BgLogger().Warn(fmt.Sprintf("Update status addr of %v from %v to %v", key, origin, httpAddr))
+		err := ddlutil.PutKVToEtcd(d.ctx, d.etcdCli, 1, key, httpAddr)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+
+	return nil
+}
+
+func updateTiFlashStores(tikvHelper *helper.Helper, pollTiFlashContext *TiFlashManagementContext) error {
+	// We need the up-to-date information about TiFlash stores.
+	// Since TiFlash Replica synchronize may happen immediately after new TiFlash stores are added.
+	tikvStats, err := tikvHelper.GetStoresStat()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	pollTiFlashContext.TiFlashStores = make(map[int64]helper.StoreStat)
+	for _, store := range tikvStats.Stores {
+		for _, l := range store.Store.Labels {
+			if l.Key == "engine" && l.Value == "tiflash" {
+				pollTiFlashContext.TiFlashStores[store.Store.ID] = store
+				logutil.BgLogger().Debug("Find tiflash store", zap.Int64("id", store.Store.ID), zap.String("Address", store.Store.Address), zap.String("StatusAddress", store.Store.StatusAddress))
+			}
+		}
+	}
+	logutil.BgLogger().Info("updateTiFlashStores finished", zap.Int("count", len(pollTiFlashContext.TiFlashStores)))
+	return nil
+}
+
+func (d *ddl) PollTiFlashReplicaStatus(ctx sessionctx.Context, pollTiFlashContext *TiFlashManagementContext) (bool, error) {
```

Comment:
If you don't invoke it in other interface, let us make it private, same to other functions.

Reply:
fixed